File: suc/kotlin/039.kt
package com.example.singlefileui

import android.os.Bundle
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import kotlin.math.sin
import kotlin.math.PI

private const val NAME = "051*InstallmentPlan*en"
private const val UI_TYPE = "Finance"
private const val STYLE_THEME = "Warm Gradient"
private const val LANG = "en"
private const val BASELINE_SIZE = "720x1280"

object AppTokens {
    object Colors {
        val primary = Color(0xFFFFA500)
        val secondary = Color(0xFFFFC107)
        val tertiary = Color(0xFFFFE066)
        val background = Color(0xFFFFFBF2)
        val surface = Color(0xFFFFFFFF)
        val surfaceVariant = Color(0xFFFFEED6)
        val outline = Color(0xFFF1C27D)
        val success = Color(0xFF22C55E)
        val warning = Color(0xFFF59E0B)
        val error = Color(0xFFEF4444)
        val onPrimary = Color(0xFFFFFFFF)
        val onSecondary = Color(0xFF1E1E1E)
        val onTertiary = Color(0xFF1E1E1E)
        val onBackground = Color(0xFF1E1E1E)
        val onSurface = Color(0xFF1E1E1E)
    }
    object TypographyTokens {
        val display = TextStyle(fontSize = 28.sp, fontWeight = FontWeight.Bold)
        val title = TextStyle(fontSize = 18.sp, fontWeight = FontWeight.Medium)
        val body = TextStyle(fontSize = 14.sp, fontWeight = FontWeight.Normal)
        val label = TextStyle(fontSize = 12.sp, fontWeight = FontWeight.Medium)
    }
    object Shapes {
        val small = RoundedCornerShape(6.dp)
        val medium = RoundedCornerShape(12.dp)
        val large = RoundedCornerShape(20.dp)
    }
    object Spacing {
        val sm = 8.dp
        val md = 12.dp
        val lg = 16.dp
        val xl = 24.dp
        val xxl = 32.dp
    }
    data class ShadowSpec(val elevation: Dp, val radius: Dp, val dy: Dp, val opacity: Float)
    object ElevationMapping {
        val level1 = ShadowSpec(2.dp, 4.dp, 2.dp, 0.12f)
        val level2 = ShadowSpec(6.dp, 8.dp, 4.dp, 0.18f)
    }
}

private val AppColorScheme = lightColorScheme(
    primary = AppTokens.Colors.primary,
    onPrimary = AppTokens.Colors.onPrimary,
    secondary = AppTokens.Colors.secondary,
    onSecondary = AppTokens.Colors.onSecondary,
    tertiary = AppTokens.Colors.tertiary,
    onTertiary = AppTokens.Colors.onTertiary,
    background = AppTokens.Colors.background,
    onBackground = AppTokens.Colors.onBackground,
    surface = AppTokens.Colors.surface,
    onSurface = AppTokens.Colors.onSurface,
    surfaceVariant = AppTokens.Colors.surfaceVariant,
    outline = AppTokens.Colors.outline,
    error = AppTokens.Colors.error
)

private val AppTypography = Typography(
    displayLarge = AppTokens.TypographyTokens.display,
    titleMedium = AppTokens.TypographyTokens.title,
    bodyMedium = AppTokens.TypographyTokens.body,
    labelMedium = AppTokens.TypographyTokens.label
)

@Composable
fun AppTheme(content: @Composable () -> Unit) {
    MaterialTheme(
        colorScheme = AppColorScheme,
        typography = AppTypography,
        shapes = Shapes(
            small = AppTokens.Shapes.small,
            medium = AppTokens.Shapes.medium,
            large = AppTokens.Shapes.large
        ),
        content = content
    )
}

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RootScreen() {
    var paymentProgress by remember { mutableFloatStateOf(0.3f) }

    Scaffold(
        contentWindowInsets = WindowInsets(0),
        topBar = {
            CenterAlignedTopAppBar(
                title = { Text("Installment Plan", style = MaterialTheme.typography.displayLarge, color = AppTokens.Colors.onSurface) },
                colors = TopAppBarDefaults.centerAlignedTopAppBarColors(containerColor = AppTokens.Colors.background)
            )
        },
        containerColor = AppTokens.Colors.background
    ) { pad ->
        Column(
            modifier = Modifier
                .fillMaxSize()
                .padding(pad)
                .background(
                    Brush.verticalGradient(
                        listOf(
                            AppTokens.Colors.secondary.copy(alpha = 0.25f),
                            AppTokens.Colors.background,
                            AppTokens.Colors.primary.copy(alpha = 0.25f)
                        )
                    )
                )
                .padding(AppTokens.Spacing.lg),
            verticalArrangement = Arrangement.spacedBy(AppTokens.Spacing.lg),
            horizontalAlignment = Alignment.CenterHorizontally
        ) {
            Surface(
                shape = AppTokens.Shapes.large,
                color = AppTokens.Colors.surface.copy(alpha = 0.9f),
                tonalElevation = AppTokens.ElevationMapping.level2.elevation,
                modifier = Modifier.fillMaxWidth().height(240.dp)
            ) {
                Canvas(modifier = Modifier.fillMaxSize().padding(AppTokens.Spacing.lg)) {
                    val w = size.width
                    val h = size.height
                    val step = w / 20
                    var prevY = h / 2
                    for (i in 0..20) {
                        val x = i * step
                        val y = h / 2 + sin(i * PI / 5 + paymentProgress * PI).toFloat() * (h / 3)
                        drawLine(
                            color = AppTokens.Colors.primary,
                            start = Offset(x - step, prevY),
                            end = Offset(x, y),
                            strokeWidth = 5f
                        )
                        prevY = y
                    }
                }
            }
            Text("Payment Progress", style = MaterialTheme.typography.titleMedium, color = AppTokens.Colors.primary)
            LinearProgressIndicator(
                progress = { paymentProgress },
                color = AppTokens.Colors.primary,
                trackColor = AppTokens.Colors.surfaceVariant,
                modifier = Modifier.fillMaxWidth().height(8.dp)
            )
            Slider(
                value = paymentProgress,
                onValueChange = { paymentProgress = it },
                colors = SliderDefaults.colors(
                    thumbColor = AppTokens.Colors.secondary,
                    activeTrackColor = AppTokens.Colors.primary,
                    inactiveTrackColor = AppTokens.Colors.surfaceVariant
                ),
                modifier = Modifier.fillMaxWidth()
            )
            Button(
                onClick = { paymentProgress = 0f },
                shape = AppTokens.Shapes.medium,
                colors = ButtonDefaults.buttonColors(containerColor = AppTokens.Colors.tertiary, contentColor = AppTokens.Colors.onTertiary),
                modifier = Modifier.fillMaxWidth().height(48.dp)
            ) {
                Text("Reset Progress", style = MaterialTheme.typography.titleMedium)
            }
        }
    }

}

class MainActivity : ComponentActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        WindowCompat.setDecorFitsSystemWindows(window, false)
        val controller = WindowInsetsControllerCompat(window, window.decorView)
        controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
        controller.hide(WindowInsetsCompat.Type.systemBars())
        setContent {
            AppTheme {
                Surface(color = MaterialTheme.colorScheme.background) {
                    RootScreen()
                }
            }
        }
    }
    override fun onWindowFocusChanged(hasFocus: Boolean) {
        super.onWindowFocusChanged(hasFocus)
        if (hasFocus) {
            val controller = WindowInsetsControllerCompat(window, window.decorView)
            controller.systemBarsBehavior = WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            controller.hide(WindowInsetsCompat.Type.systemBars())
        }
    }
}

@Preview(showBackground = true, backgroundColor = 0xFFFFFBF2)
@Composable
fun PreviewScreen() {
    AppTheme {
        RootScreen()
    }
}

File: suc/swift/039.swift
import SwiftUI
import CoreGraphics // For CGFloat and sin
import UIKit      // For UISlider (via UIViewRepresentable)

// MARK: - AppTokens
// 统一的 UI 令牌，方便修改颜色、字体、间距和形状
struct AppTokens {
    struct Colors {
        static let primary = Color(red: 0xFF / 255.0, green: 0xA5 / 255.0, blue: 0x00 / 255.0) // #FFA500
        static let secondary = Color(red: 0xFF / 255.0, green: 0xC1 / 255.0, blue: 0x07 / 255.0) // #FFC107
        static let tertiary = Color(red: 0xFF / 255.0, green: 0xE0 / 255.0, blue: 0x66 / 255.0) // #FFE066
        static let background = Color(red: 0xFF / 255.0, green: 0xFB / 255.0, blue: 0xF2 / 255.0) // #FFFBF2
        static let surface = Color(red: 0xFF / 255.0, green: 0xFF / 255.0, blue: 0xFF / 255.0) // #FFFFFF
        static let surfaceVariant = Color(red: 0xFF / 255.0, green: 0xEE / 255.0, blue: 0xD6 / 255.0) // #FFEED6
        static let outline = Color(red: 0xF1 / 255.0, green: 0xC2 / 255.0, blue: 0x7D / 255.0) // #F1C27D
        static let success = Color(red: 0x22 / 255.0, green: 0xC5 / 255.0, blue: 0x5E / 255.0) // #22C55E
        static let warning = Color(red: 0xF5 / 255.0, green: 0x9E / 255.0, blue: 0x0B / 255.0) // #F59E0B
        static let error = Color(red: 0xEF / 255.0, green: 0x44 / 255.0, blue: 0x44 / 255.0) // #EF4444
        static let onPrimary = Color(red: 0xFF / 255.0, green: 0xFF / 255.0, blue: 0xFF / 255.0) // #FFFFFF
        static let onSecondary = Color(red: 0x1E / 255.0, green: 0x1E / 255.0, blue: 0x1E / 255.0) // #1E1E1E
        static let onTertiary = Color(red: 0x1E / 255.0, green: 0x1E / 255.0, blue: 0x1E / 255.0) // #1E1E1E
        static let onBackground = Color(red: 0x1E / 255.0, green: 0x1E / 255.0, blue: 0x1E / 255.0) // #1E1E1E
        static let onSurface = Color(red: 0x1E / 255.0, green: 0x1E / 255.0, blue: 0x1E / 255.0) // #1E1E1E
    }

    struct TypographyTokens {
        static let display = Font.system(size: 28, weight: .bold)
        static let title = Font.system(size: 18, weight: .medium)
        static let body = Font.system(size: 14, weight: .regular) // Kotlin's FontWeight.Normal maps to .regular
        static let label = Font.system(size: 12, weight: .medium)
    }

    struct Shapes {
        static let small: CGFloat = 6.0
        static let medium: CGFloat = 12.0
        static let large: CGFloat = 20.0
    }

    struct Spacing {
        static let sm: CGFloat = 8.0
        static let md: CGFloat = 12.0
        static let lg: CGFloat = 16.0
        static let xl: CGFloat = 24.0
        static let xxl: CGFloat = 32.0
    }

    // Compose ShadowSpec(elevation: Dp, radius: Dp, dy: Dp, opacity: Float)
    // SwiftUI shadow(color: Color, radius: CGFloat, x: CGFloat, y: CGFloat)
    // We map Compose's `radius` to SwiftUI's `radius` (blur radius) and `dy` to `y` offset.
    struct ShadowSpec {
        let radius: CGFloat // Blur radius
        let dy: CGFloat     // Y offset
        let opacity: Double // Shadow color opacity
    }

    struct ElevationMapping {
        static let level1 = ShadowSpec(radius: 4, dy: 2, opacity: 0.12)
        static let level2 = ShadowSpec(radius: 8, dy: 4, opacity: 0.18)
    }
}

// MARK: - Custom SwiftUI Components

/// 自定义线性进度指示器，以匹配 Compose 版本的颜色和高度
struct CustomLinearProgressIndicator: View {
    let progress: Double
    let color: Color
    let trackColor: Color
    let height: CGFloat

    var body: some View {
        GeometryReader { geometry in
            ZStack(alignment: .leading) {
                trackColor // 背景轨道
                    .frame(width: geometry.size.width, height: height)
                    .cornerRadius(height / 2)

                color // 进度条
                    .frame(width: geometry.size.width * CGFloat(progress), height: height)
                    .cornerRadius(height / 2)
            }
        }
        .frame(height: height)
    }
}

/// 自定义滑块，使用 UIViewRepresentable 封装 UISlider，以在 iOS 16.0 上精确控制滑块拇指颜色
struct CustomSlider: UIViewRepresentable {
    @Binding var value: Float
    let activeTrackColor: Color
    let inactiveTrackColor: Color
    let thumbColor: Color

    func makeUIView(context: Context) -> UISlider {
        let slider = UISlider()
        slider.minimumValue = 0
        slider.maximumValue = 1
        slider.value = value
        slider.addTarget(context.coordinator, action: #selector(Coordinator.valueChanged(_:)), for: .valueChanged)

        // 设置轨道颜色
        slider.minimumTrackTintColor = UIColor(activeTrackColor)
        slider.maximumTrackTintColor = UIColor(inactiveTrackColor)

        // 设置滑块拇指颜色
        slider.thumbTintColor = UIColor(thumbColor)

        return slider
    }

    func updateUIView(_ uiView: UISlider, context: Context) {
        uiView.value = value
        uiView.minimumTrackTintColor = UIColor(activeTrackColor)
        uiView.maximumTrackTintColor = UIColor(inactiveTrackColor)
        uiView.thumbTintColor = UIColor(thumbColor)
    }

    func makeCoordinator() -> Coordinator {
        Coordinator(self)
    }

    class Coordinator: NSObject {
        var parent: CustomSlider

        init(_ parent: CustomSlider) {
            self.parent = parent
        }

        @objc func valueChanged(_ sender: UISlider) {
            parent.value = sender.value
        }
    }
}

// MARK: - Helper Extensions
extension Double {
    /// 将 Double 转换为 CGFloat，用于绘图
    func toCGFloat() -> CGFloat {
        return CGFloat(self)
    }
}

// MARK: - Main UI Components

/// 包含 Canvas 绘图的 Surface 区域
struct SurfaceWithCanvas: View {
    @Binding var paymentProgress: Float

    var body: some View {
        ZStack {
            // Surface 背景和阴影
            RoundedRectangle(cornerRadius: AppTokens.Shapes.large)
                .fill(AppTokens.Colors.surface.opacity(0.9))
                .shadow(color: Color.black.opacity(AppTokens.ElevationMapping.level2.opacity),
                        radius: AppTokens.ElevationMapping.level2.radius,
                        x: 0,
                        y: AppTokens.ElevationMapping.level2.dy)
                .frame(height: 240) // 固定高度，与 Kotlin 版本一致

            // Canvas 绘图区域
            Canvas { context, size in
                let w = size.width
                let h = size.height
                let step = w / 20 // 20 个步长

                var path = Path()
                
                // 计算第一个点 (x=0) 的 Y 坐标，并移动到该点
                let initialY = h / 2 + sin(0 * Double.pi / 5 + Double(paymentProgress) * Double.pi).toCGFloat() * (h / 3)
                path.move(to: CGPoint(x: 0, y: initialY))

                // 绘制后续的线段，从 i=1 开始
                for i in 1...20 {
                    let x = CGFloat(i) * step
                    let y = h / 2 + sin(Double(i) * Double.pi / 5 + Double(paymentProgress) * Double.pi).toCGFloat() * (h / 3)
                    path.addLine(to: CGPoint(x: x, y: y))
                }
                context.stroke(path, with: .color(AppTokens.Colors.primary), lineWidth: 5)
            }
            .padding(AppTokens.Spacing.lg) // Canvas 内部的内边距
            .frame(height: 240) // 确保 Canvas 占据其父 Surface 的全部高度
        }
        .frame(maxWidth: .infinity) // 填充父 VStack 的宽度
    }
}

// MARK: - Root Screen
/// 应用程序的根视图，包含所有 UI 元素和布局
struct RootScreen: View {
    @State private var paymentProgress: Float = 0.3 // 支付进度状态

    var body: some View {
        // 模拟 Compose Scaffold 的结构
        VStack(spacing: 0) { // 最外层 VStack，用于放置 Top Bar 和内容区域，间距为 0
            // 顶部栏 (对应 Compose 的 CenterAlignedTopAppBar)
            Text("Installment Plan")
                .font(AppTokens.TypographyTokens.display)
                .foregroundColor(AppTokens.Colors.onSurface)
                .frame(maxWidth: .infinity)
                .frame(height: 64) // Compose Material3 TopAppBar 的默认高度为 64.dp
                .background(AppTokens.Colors.background) // 顶部栏背景色

            // 主要内容区域 (对应 Compose Scaffold 的 content lambda)
            VStack(spacing: AppTokens.Spacing.lg) { // 内容区域内部的元素间距
                SurfaceWithCanvas(paymentProgress: $paymentProgress)

                Text("Payment Progress")
                    .font(AppTokens.TypographyTokens.title)
                    .foregroundColor(AppTokens.Colors.primary)

                CustomLinearProgressIndicator(
                    progress: Double(paymentProgress),
                    color: AppTokens.Colors.primary,
                    trackColor: AppTokens.Colors.surfaceVariant,
                    height: 8
                )
                .frame(maxWidth: .infinity) // 填充宽度

                CustomSlider(
                    value: $paymentProgress,
                    activeTrackColor: AppTokens.Colors.primary,
                    inactiveTrackColor: AppTokens.Colors.surfaceVariant,
                    thumbColor: AppTokens.Colors.secondary
                )
                .frame(maxWidth: .infinity) // 填充宽度

                Button(action: {
                    paymentProgress = 0.0 // 重置进度
                }) {
                    Text("Reset Progress")
                        .font(AppTokens.TypographyTokens.title)
                        .foregroundColor(AppTokens.Colors.onTertiary)
                        .frame(maxWidth: .infinity, maxHeight: .infinity) // 使文本填充按钮
                }
                .frame(height: 48) // 按钮高度
                .background(AppTokens.Colors.tertiary) // 按钮背景色
                .cornerRadius(AppTokens.Shapes.medium) // 按钮圆角
                .buttonStyle(PlainButtonStyle()) // 移除 SwiftUI 默认按钮样式，以便自定义背景和圆角
            }
            .padding(AppTokens.Spacing.lg) // 内容区域的内边距
            .frame(maxWidth: .infinity, maxHeight: .infinity) // 填充剩余空间
            .background(
                LinearGradient( // 内容区域的渐变背景
                    gradient: Gradient(colors: [
                        AppTokens.Colors.secondary.opacity(0.25),
                        AppTokens.Colors.background,
                        AppTokens.Colors.primary.opacity(0.25)
                    ]),
                    startPoint: .top,
                    endPoint: .bottom
                )
            )
        }
        .background(AppTokens.Colors.background) // 整个屏幕的背景色 (Scaffold containerColor)
        .ignoresSafeArea() // 使视图忽略安全区域，实现全屏
        .statusBarHidden(true) // 隐藏顶部状态栏
    }
}

// MARK: - App Entry Point
/// 应用程序的入口点
@main
struct InstallmentPlanApp: App {
    var body: some Scene {
        WindowGroup {
            RootScreen()
        }
    }
}